mongomock/store.py

```diff
@@ -1,11 +1,91 @@
 import collections
 import datetime
+from contextlib import contextmanager
+
 import mongomock  # Used for utcnow - please see https://github.com/mongomock/mongomock#utcnow
 import six
 import six.moves
 import threading
 
-lock = threading.RLock()
+
+class RWLock:
+    """Lock for enabling multiple readers but only 1 exclusive writer
+    to access a resource
+
+    Source: https://cutt.ly/Ij70qaq
+    """
+
+    def __init__(self):
+        self._read_switch = _LightSwitch()
+        self._write_switch = _LightSwitch()
+        self._no_readers = threading.Lock()
```

Comment:
I suggest you use RLock instead Lock so that a thread can actually use it multiple times.

Reply:
It seems to be already covered. For instance, it's possible to do:
```
lock = RWLock()
with lock.reader():
    with lock.reader():
        ...
```
And it doesn't block the thread. Is that what you meant? 

I now added the `RLock` to the `_readers_queue`. Other than that it changes the behavior (exceptions, deadlocks, broken tests and panic lol).

Reply:
This implementation from [wikipedia](https://en.wikipedia.org/wiki/Readers%E2%80%93writers_problem#Third_readers%E2%80%93writers_problem) might be cleaner (and more fair):
```python
import threading
from contextlib import contextmanager

import threading


class RWLock:

    def __init__(self):
        self.readcount = 0
        # Controls the access to readcount
        self.readcount_lock = threading.Lock()

        # Controls the access to the critical region
        self.resource = threading.Lock()

        # Queue
        self.servicequeue = threading.RLock()

    @contextmanager
    def reader(self):
        self._acquire_for_reader()
        try:
            yield
        except Exception:
            raise
        finally:
            self._release_for_reader()

    def _acquire_for_reader(self):
        # Get in line
        self.servicequeue.acquire()
        # Lock for updating active readers
        self.readcount_lock.acquire()

        self.readcount += 1
        if self.readcount == 1:
            # First reader locks the resource for other writers
            self.resource.acquire()

        # Allow next reader/writer
        self.servicequeue.release()
        # Release access to readcount
        self.readcount_lock.release()

    def _release_for_reader(self):
        # Get access to readcount
        self.readcount_lock.acquire()
        self.readcount -= 1
        if self.readcount == 0:
            # If no one else is reading, release the resource
            self.resource.release()
        # Release access to readcount
        self.readcount_lock.release()

    @contextmanager
    def writer(self):
        self._acquire_for_writer()
        try:
            yield
        except Exception:
            raise
        finally:
            self._release_for_writer()

    def _acquire_for_writer(self):
        # Get in line
        self.servicequeue.acquire()
        # Request resource
        self.resource.acquire()
        # Let next in line to be serviced
        self.servicequeue.release()

    def _release_for_writer(self):
        # Release for next reader/writer
        self.resource.release()
```